### PR TITLE
feat(rdb): add password generator

### PR DESCRIPTION
**Pull request by yfodil**
feat #1155

### CODE REVIEW

File: internal/namespaces/rdb/v1/custom_instance.go [hunk 1]
```diff
@@ -109,24 +110,171 @@
...
+			createInstanceRequest := customRequest.CreateInstanceRequest
+
+			client := core.ExtractClient(ctx)
+			api := rdb.NewAPI(client)
+
+			createInstanceRequest.NodeType = strings.ToLower(createInstanceRequest.NodeType)
+
+			if customRequest.GeneratePassword {
+				createInstanceRequest.Password = passwordgenerator.GeneratePassword(21, 1, 1, 1, 1)
+				fmt.Printf("Your generated password is %s \n", createInstanceRequest.Password)
```
[the remark this answers is not in the document]
Comment (yfodil) [reply internal/namespaces/rdb/v1/custom_instance.go:259]: Wanted behavior

File: internal/namespaces/rdb/v1/custom_user.go [hunk 2]
```diff
@@ -75,3 +78,149 @@
...
+			customRequest := args.(*rdbCreateUserRequestCustom)
+
+			createUserRequest := customRequest.CreateUserRequest
+
+			client := core.ExtractClient(ctx)
+			api := rdb.NewAPI(client)
+
+			if customRequest.GeneratePassword {
+				createUserRequest.Password = passwordgenerator.GeneratePassword(21, 1, 1, 1, 1)
+				fmt.Printf("Your generated password is %s \n", createUserRequest.Password)
```
[the remark this answers is not in the document]
Comment (yfodil) [reply internal/namespaces/rdb/v1/custom_user.go:146]: Wanted behavior

File: internal/passwordgenerator/password_generator.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	lowerLetters   = "abcdedfghijklmnopqrstuvwxyz"
+	upperLetters   = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	specialSymbols = "!$%^&*()_+{}:@[];'#<>?,./|\\\\-=?"
+	allSet         = lowerLetters + upperLetters + specialSymbols + numbers
+)
+
+func GeneratePassword(length, minNumbers, minLower, minUpper, minSymbol int) string {
+	var password strings.Builder
+
+	rand.Seed(time.Now().Unix())
```
[the remark this answers is not in the document]
Comment (yfodil) [reply internal/passwordgenerator/password_generator.go:22]: Used just for the final shuffle

File: internal/namespaces/rdb/v1/testdata/test-update-user-simple.golden [hunk 4]
```diff
@@ -0,0 +1,24 @@
+🎲🎲🎲 EXIT CODE: 1 🎲🎲🎲
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/testdata/test-update-user-simple.golden:1]: Why does this golden is updated?

File: internal/namespaces/rdb/v1/testdata/test-update-user-with-password-generator.golden [hunk 5]
```diff
@@ -0,0 +1,24 @@
+🎲🎲🎲 EXIT CODE: 1 🎲🎲🎲
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/testdata/test-update-user-with-password-generator.golden:1]: Why is the golden like this?

File: internal/passwordgenerator/password_generator.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+		random := randInt(len(allSet))
+		password.WriteString(string(allSet[random]))
+	}
+	inRune := []rune(password.String())
+	for i := range inRune {
+		j := randInt(i + 1)
+		inRune[i], inRune[j] = inRune[j], inRune[i]
+	}
+
+	return string(inRune)
```
Comment (remyleone) [inline internal/passwordgenerator/password_generator.go:51]: Are all Numbers, upper and lower always returned in the same order?

File: internal/namespaces/rdb/v1/testdata/test-create-user-with-password-generator.golden [hunk 7]
```diff
@@ -0,0 +1,9 @@
+🎲🎲🎲 EXIT CODE: 0 🎲🎲🎲
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/testdata/test-create-user-with-password-generator.golden:1]: Where is the password showing up here?

File: internal/namespaces/rdb/v1/custom_user.go [hunk 8]
```diff
@@ -75,3 +78,149 @@ func userListBuilder(c *core.Command) *core.Command {
...
+		GeneratePassword bool
+	}
+
+	return &core.Command{
+		Short:     `Create a user on a given instance`,
+		Long:      `Create a user on a given instance.`,
+		Namespace: "rdb",
+		Resource:  "user",
+		Verb:      "create",
+		// Deprecated:    false,
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/custom_user.go:94]: Remove this commented line

File: internal/namespaces/rdb/v1/custom_user.go [hunk 9]
```diff
@@ -75,3 +78,149 @@ func userListBuilder(c *core.Command) *core.Command {
...
+				Name:       "password",
+				Short:      `Password of the user you want to create`,
+				Required:   false,
+				Deprecated: false,
+				Positional: false,
+				OneOfGroup: "password",
+			},
+			{
+				Name:       "generate-password",
+				Short:      `Will generate a valid password for the user`,
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/custom_user.go:121]: Maybe indicate what is considered as a valid password

File: internal/namespaces/rdb/v1/custom_user.go [hunk 10]
```diff
@@ -75,3 +78,149 @@ func userListBuilder(c *core.Command) *core.Command {
...
+				Required:   false,
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/custom_user.go:122]: Should we put this field as a default true?

File: internal/namespaces/rdb/v1/custom_user.go [hunk 11]
```diff
@@ -75,3 +78,149 @@
...
+
+			updateUserRequest := customRequest.UpdateUserRequest
+
+			client := core.ExtractClient(ctx)
+			api := rdb.NewAPI(client)
+
+			if customRequest.GeneratePassword {
+				updateUserRequest.Password = new(string)
+				*updateUserRequest.Password = passwordgenerator.GeneratePassword(21, 1, 1, 1, 1)
+				fmt.Printf("Your generated password is %v \n", *updateUserRequest.Password)
```
[the remark this answers is not in the document]
Comment (remyleone) [reply internal/namespaces/rdb/v1/custom_user.go:219]: I wonder how we can turn off CodeQL behavior. Maybe we can exclude CodeQL from this code path.

File: internal/namespaces/rdb/v1/custom_user.go (hunk 10 above)
Comment (yfodil) [reply internal/namespaces/rdb/v1/custom_user.go:125]: Yes

File: internal/passwordgenerator/password_generator.go (hunk 6 above)
Comment (yfodil) [reply internal/passwordgenerator/password_generator.go:51]: Yes this part does the shuffle: 
```
inRune := []rune(password.String())
	for i := range inRune {
		j := randInt(i + 1)
		inRune[i], inRune[j] = inRune[j], inRune[i]
	}
```

File: internal/namespaces/rdb/v1/testdata/test-update-instance-set-a-timezone.golden [hunk 12]
```diff
@@ -1,25 +1,27 @@
 🎲🎲🎲 EXIT CODE: 0 🎲🎲🎲
 🟩🟩🟩 STDOUT️ 🟩🟩🟩️
+
```
[the remark this answers is not in the document]
Comment (yfodil) [reply internal/namespaces/rdb/v1/testdata/test-update-instance-set-a-timezone.golden:3]: I introduced blank lines at the top and the bottom, in the `instanceMarshalerFunc`, this helps in `rdb instance create` command to separate instance info and the password 

The little drowback is that it adds superfluous blank lines too whenever an rdb instance object is returned, like here

File: internal/namespaces/rdb/v1/testdata/test-update-instance-modify3-settings+-add-a-new-one.golden [hunk 13]
```diff
@@ -2,24 +2,24 @@
 🟩🟩🟩 STDOUT️ 🟩🟩🟩️
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/testdata/test-update-instance-modify3-settings+-add-a-new-one.golden:2]: Why does this golden requires a change?

File: internal/namespaces/rdb/v1/testdata/test-update-instance-set-a-timezone.golden [hunk 14]
```diff
@@ -2,24 +2,24 @@
 🟩🟩🟩 STDOUT️ 🟩🟩🟩️
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/testdata/test-update-instance-set-a-timezone.golden:2]: Why does this requires a change?

File: internal/namespaces/rdb/v1/testdata/test-update-instance-update-instance-name.golden [hunk 15]
```diff
@@ -2,24 +2,24 @@
 🟩🟩🟩 STDOUT️ 🟩🟩🟩️
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/testdata/test-update-instance-update-instance-name.golden:2]: Why is this change required?

File: internal/namespaces/rdb/v1/testdata/test-update-user-simple.golden [hunk 16]
```diff
@@ -0,0 +1,11 @@
+🎲🎲🎲 EXIT CODE: 0 🎲🎲🎲
+🟩🟩🟩 STDOUT️ 🟩🟩🟩️
+Name      cli-test
+IsAdmin   true
+Password  Newp1ssw0rd!
+🟩🟩🟩 JSON STDOUT 🟩🟩🟩
+{
+  "name": "cli-test",
+  "is_admin": true,
+  "Password": "Newp1ssw0rd!"
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/testdata/test-update-user-simple.golden:10]: use password to stay consistent with the case of the other component of the JSON

File: internal/namespaces/rdb/v1/testdata/test-upgrade-instance-simple.golden [hunk 17]
```diff
@@ -2,24 +2,24 @@
 🟩🟩🟩 STDOUT️ 🟩🟩🟩️
 CreatedAt                   few seconds ago
 Region                      fr-par
-ID                          143f084f-2d48-48d2-9094-4363c81d6fcf
+ID                          d30cc9a2-e9fa-4ff0-81e1-80c661fd9e2c
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/testdata/test-upgrade-instance-simple.golden:5]: Why is this change required?

File: internal/namespaces/rdb/v1/testdata/test-get-instance-simple.golden [hunk 18]
```diff
@@ -2,24 +2,24 @@
 🟩🟩🟩 STDOUT️ 🟩🟩🟩️
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/testdata/test-get-instance-simple.golden:2]: Why is this change required?

File: internal/namespaces/rdb/v1/custom_acl_test.go [hunk 19]
```diff
@@ -10,7 +10,7 @@ func Test_AddACL(t *testing.T) {
 	t.Run("Simple", core.Test(&core.TestConfig{
 		Commands:   GetCommands(),
 		BeforeFunc: createInstance("PostgreSQL-12"),
-		Cmd:        "scw rdb acl add instance-id={{ .Instance.ID }} rules.0.ip=4.2.3.4",
+		Cmd:        "scw rdb acl add instance-id={{ $.Instance.Instance.ID }} rules.0.ip=4.2.3.4",
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/custom_acl_test.go:13]: Why is this change required?

File: internal/namespaces/rdb/v1/custom_backup_test.go [hunk 20]
```diff
@@ -22,7 +22,7 @@ func Test_CreateBackup(t *testing.T) {
 				},
 			),
 		),
-		Cmd:           "scw rdb backup create name=foobar expires-at=2999-01-02T15:04:05-07:00 instance-id={{ .Instance.ID }} database-name=rdb --wait",
+		Cmd:           "scw rdb backup create name=foobar expires-at=2999-01-02T15:04:05-07:00 instance-id={{ $.Instance.Instance.ID }} database-name=rdb --wait",
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/custom_backup_test.go:25]: Why is this change required?

File: internal/namespaces/rdb/v1/custom_instance.go [hunk 21]
```diff
@@ -73,6 +75,42 @@ func instanceMarshalerFunc(i interface{}, opt *human.MarshalOpt) (string, error)
 	return str, nil
 }
 
+type createInstanceResult struct {
+	Instance *rdb.Instance
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/custom_instance.go:79]: Why do you need to have Instance named? Is it possible to just have:

```go
type createInstanceResult struct {
	*rdb.Instance
	Password string
}
```

File: internal/namespaces/rdb/v1/custom_instance_test.go [hunk 22]
```diff
@@ -33,16 +33,23 @@ func Test_CreateInstance(t *testing.T) {
 	t.Run("Simple", core.Test(&core.TestConfig{
 		Commands:  GetCommands(),
 		Cmd:       fmt.Sprintf("scw rdb instance create node-type=DB-DEV-S is-ha-cluster=false name=%s engine=%s user-name=%s password=%s --wait", name, engine, user, password),
-		Check:     core.TestCheckGolden(),
-		AfterFunc: core.ExecAfterCmd("scw rdb instance delete {{ .CmdResult.ID }}"),
+		Check:     core.TestCheckExitCode(0),
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/custom_instance_test.go:36]: Why there is not a check golden?

File: internal/namespaces/rdb/v1/testdata/test-clone-instance-simple.golden [hunk 23]
```diff
@@ -2,24 +2,24 @@
 🟩🟩🟩 STDOUT️ 🟩🟩🟩️
 CreatedAt                   few seconds ago
 Region                      fr-par
-ID                          1221fbbf-b6b8-48ba-a337-f9df98c8e2a6
+ID                          28fea723-ef0a-4d19-a939-ec5d2ea0b9b2
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/testdata/test-clone-instance-simple.golden:5]: Why is this change required?

File: internal/namespaces/rdb/v1/custom_user_test.go [hunk 24]
```diff
@@ -10,8 +11,50 @@ func Test_ListUser(t *testing.T) {
 	t.Run("Simple", core.Test(&core.TestConfig{
 		Commands:   GetCommands(),
 		BeforeFunc: createInstance("PostgreSQL-12"),
-		Cmd:        "scw rdb user list instance-id={{ .Instance.ID }}",
-		Check:      core.TestCheckGolden(),
+		Cmd:        "scw rdb user list instance-id={{ $.Instance.Instance.ID }}",
+		Check:      core.TestCheckExitCode(0),
```
Comment (remyleone) [inline internal/namespaces/rdb/v1/custom_user_test.go:15]: Why is the check change required?

File: internal/namespaces/rdb/v1/custom_user.go [hunk 25]
```diff
@@ -80,3 +83,110 @@
...
+
+		customRequest := argsI.(*rdbUpdateUserRequestCustom)
+
+		updateUserRequest := customRequest.UpdateUserRequest
+
+		var err error
+		if customRequest.GeneratePassword && customRequest.Password == nil {
+			updateUserRequest.Password = new(string)
+			*updateUserRequest.Password, err = passwordgenerator.GeneratePassword(21, 1, 1, 1, 1)
+			fmt.Printf("Your generated password is %v \n", *updateUserRequest.Password)
```
[the remark this answers is not in the document]
Comment (yfodil) [reply internal/namespaces/rdb/v1/custom_user.go:177]: Wanted behavior

**Review by remyleone — APPROVED** [review]
